[28기 박준영] 로그인, 메인 페이지 레이아웃, 로직 완성 및 추가 기능 구현

joonyg10 opened the pull request:
- Main페이지 레이아웃 작성중
- Main페이지 feed 이미지 슬라이더 완성
- Main페이지 댓글 달기 로직 작성중

## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 리뷰 반영
- [x] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
1. 아이디, 비밀번호 입력창을 사용한 로그인 기능 구현
   - 아이디와 비밀번호를 조건에 맞게 입력 시 버튼이 활성화 되며 main페이지로 이동하도록 한다.
   
2. Main페이지 레이아웃 작성
   - Main페이지의 스토리 캐러셀 왼쪽, 오른쪽 버튼 클릭 시 슬라이드 되도록 구현
   - feed의 이미지 슬라이드 구현
   - feed 이미지 슬라이드 밑의 navigation 버튼 구현 및 이미지 슬라이드와 링크해 움직이도록 구현
   - feed의 각 댓글을 왼쪽으로 드래그 하면 X (드래그 취소)버튼, 삭제(해당 댓글 삭제) 버튼 구현
   - feed 하단에 댓글을 작성하고 업로드 되도록 구현
   
<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. 로그인 버튼 활성화
   - ID(@포함)와 password(6자 이상)을 작성한 경우에 버튼에 active라는 클래스를 주어 활성화 여부를 부여.
   - password input에서 enter 또는 로그인 버튼 클릭 시 active라는 클래스를 버튼이 가지고 있다면 main페이지로 이동되도록 구현함.

2. Main 페이지 레이아웃 작성
  - 네비게이션 바, 스토리, 사이드바, 피드로 구성된 Main 페이지 레이아웃 작성 (상세 내용은 아래 기술함)

3. 네비게이션 바
  - 검색 input에 입력 시 입력된 string을 포함하고 있는 인물의 리스트가 Input바 밑에 나타나게 구현함.
  - input에 마우스로 focus를 맞출 수 모든 인물이 나타남
  - 검색 바에서 esc를 누르거나 focusout 이벤트가 발생할 때, 리스트가 사라진다.
  - 네비게이션 바 오른쪽 상단 이미지를 클릭했을 때, 메뉴 박스가 나타난다.
  
4. 스토리 부분
  - 각 이미지마다 그라데이션이 들어간 선을 스타일링
  - 스토리 section의 양측에 있는 버튼 클릭 시 다음 스토리들이 슬라이드 되어 나타나도록 구현
  - 스토리를 슬라이드 할 때 스토리 박스 너비의 1/2 만큼 움직이도록 하였으며, 스토리의 마지막 부분에 도달할 때 멈추도록 구현
  - 양쪽 끝에 도달 했을 때, 해당방향의 버튼이 display: hidden하여 없어지도록 하였음.
  
5. 사이드바 (추천 목록)
  - position: fixed를 이용하여 메인페이지를 스크롤 하여도 오른쪽의 같은 영역에 붙어 있도록 함

6. 피드 이미지 슬라이더
  - 스토리와 마찬가지로 버튼을 클릭했을 때, 양쪽으로 슬라이드 되며 처음과 마지막 페이지에 도달 시 해당 방향의 버튼에 display: hidden을 부여
  - 이미지 슬라이더 밑에 이미지의 갯수만큼 navigator을 생성
  - navigator는 현재 슬라이더의 인덱스에 따라 몇 번째 이미지인지 표시되며 navigator의 버튼 클릭 시 해당 인덱스의 이미지로 바로 이동된다.
  
7. 피드내의 모든 하트
  - 하트 및 다른 아이콘에 hover효과로 border-color를 옅은 회색으로 바뀌도록 스타일링
  - 하트는 svg와 path태그를 통해 가져왔으며 하트 클릭시 transform: scale이 적용되어 심장이 뛰는 듯한 효과 + 색상이 빨갛게 변하는 애니메이션을 추가 
  -  하트를 클릭했을 때, 좋아요 수가 업데이트 되도록 연동시켰으며 이는 댓글의 하트를 클릭했을 때도 똑같은 로직이 적용된다.

8. 댓글
  - 7)에서 설명한대로 하트 클릭시 애니메이션 효과와 배경색상이 red로 변한다.
  - 댓글을 왼쪽으로 드래그하면 X와 쓰레기 아이콘이 포함된 숨겨진 부분이 나타난다. (각각 크기를 width: 8%로 하였다)
  - X버튼을 클릭했을 때, 드래그 하여 움직인 것을 되돌려 주는 기능을 수행한다.
  - 쓰레기 버튼은 댓글삭제 기능 로직이 들어 있으며 클릭시 해당 댓글이 삭제된다.
  - 댓글을 드래그하는 기능은 requestAnimationFrame이라는 메소드를 사용하였으며 mousedown으로 시작지점을 저장하고 mousemove이벤트로 이동한 마우스의 위치와 계산해 댓글.style.transform에 동적으로 업데이트 하였다.
  - 이 때, 이동한 거리가 특정 범위 (댓글 너비의 16%로 버튼 2개의 크기와 똑같이 설정)를 넘어 갔을 때, 넘어간 범위를 무효로 해 너비16%만 translate되도록 설정하였다.
  - 마찬가지로 작은 거리만을 드래그 하여도 아이콘들이 보이도록 설정하였다.

9. 댓글 입력 창 
  - 댓글 input 또는 '게시' 버튼을 클릭 했을 때, 작성된 댓글이 있다면 (empty string이 아니라면) 댓글 <ul>부분에 새로운 댓글 <li>을 createElement("li")를 통해 생성, 추가하도록 작성하였다.
  - 댓글 input에서 shift + enter를 클릭했을 때, 작성한 댓글이 게시되지 않고 새로운 줄을 생성하도록 구현
  - 입력한 댓글이 30자 이상 또는 shift+enter를 통해 여러 줄인 경우 댓글을 더보기...로 대체 하였고 더보기... 를 클릭시 작성했던 댓글이 나타나도록 하였다.


10. 스켈레톤 로딩 효과
   - Main 페이지에 있는 모든 이미지에 대해 로딩 초기 2초 동안 회색배경에 하얀색 배경이 지나가게 스타일링하여 마치 서버로 부터 이미지를 받고 있는 듯한 효과를 부여.
   - 2초 이후에 임시로 부여했던 회색 배경색상이 사라지며 원래의 이미지가 나타나도록 z-index 및 background를 조정.



<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
   - 여러 요소를 querySelector로 가져왔을 때, 관리 하는 방법을 찾아야 할 것 같습니다.
   - feed의 이미지 슬라이더를 구현할 떄, 이미지를 슬라이더의 너비와 맞추는데 어려움이 있었습니다.
   - 댓글을 새로 추가하는 상황에 있어서 createElement로 만든 li요소에 기존의 eventListener로는 접근을 할 수가 없어 새로 li요소를 만들 때, eventListener를 추가하는 방식으로 초기 방향을 잡았습니다. 하지만 새로 만든 댓글의 하트에도 접근을 하여 조작을 해야 되는 상황이 생겼고 코드가 너무 번잡해져 다른 방식을 찾아야 했습니다. 그래서 댓글 전체를 관리하는 ul태그에 대해서 click 이벤트를 주었고 클릭된 evt.target이 하트인지 아닌지에 따라 다른 로직을 시행하는 방식으로 바꾸었습니다. 결과적으로 새로 만든 li에 대해 일일히 이벤트리스너를 추가하지 않아도 되었고 ul태그, 즉 컨테이너에 대해서만 이벤트리스너를 추가하는 이벤트전파 형식으로 문제를 수월하게 해결할 수 있었습니다.
   - 마찬가지로 댓글을 삭제하는 상황에서 li -> section -> delete-btn으로 되어있는 상태에서 delete-btn이 포함되어 있는 li를 찾아야 하는 상황이 생겼는데 closest라는 메소드를 사용하여 가장 가까운 조상요소인 li에 쉽게 접근할 수 있었습니다.
   - 평소에 이벤트 리스너를 작성할 때, 함수명 대신 (e) => {}과 같이 화살표함수를 쓰는 버릇이 있어서 이벤트에 대해 무엇이 시행되는 것인지 코드를 다시볼 때 매우 힘들었습니다. 금요일 멘토님이 예시 코드를 보여주셨던 세션에서 여러 함수로 로직을 분리하는 방식을 볼 수 있었습니다. 이후 js리펙토링을 진행하면서 로직을 분리해 아직도 많이 복잡하지만 이전의 번잡했던 로직을 분리하는 고민하는 중입니다.
   - 그 외 사항으로 이미지를 컨테이너에 맞게 조절하는 방식과 직접 크기를 줄 때 유의할 점 등을 조사하려 합니다.
<br />

## :: 기타 질문 및 특이 사항

claire-euni commented on students/14th/bjkim/style/common.css:
```diff
@@ -0,0 +1,4 @@
+html {
+  font-size: 10px;
+  box-sizing: border-box;
+}
```
전체에 적용할 때는 * 을 사용하시면 됩니다~ `* { }`

claire-euni commented on students/14th/bjkim/style/login.css:
```diff
@@ -0,0 +1,74 @@
+body {
+  height: 100vh;
```
필요한 속성일까요?

claire-euni commented on students/14th/bjkim/style/login.css:
```diff
@@ -0,0 +1,74 @@
+}
+
+.maindish {
+  box-sizing: border-box;
+  height: 380px;
+  width: 350px;
+  border: 1px solid #dbdbdb;
+  margin: 12px 0 0 0;
+  padding: 10px 0 10px 0;
+  min-height: auto;
+  min-width: auto;
+  position: relative;
+  background: #ffffff;
+  margin-top: 150px;
```
css 속성은 관련도 있는 속성끼리 순서대로 묶어주시는게 좋습니다. 예를 들어 width, height / display, postion / border / font 등으로 작성해주세요~

claire-euni commented on students/14th/bjkim/style/main.css:
```diff
@@ -0,0 +1,199 @@
+body {
+  height: 100vh;
+  margin: 0%;
```
`margin:0`은 common.css에서 주면 좋겠죠?

claire-euni commented on students/14th/bjkim/main.html:
```diff
@@ -0,0 +1,96 @@
+    user-scalable-no, maximun-scale=1, minimun-scale=1"
+    />
+    <meta http-equiv="X-UA-Compatible" content="ie=edge" />
+    <title>repl.it</title>
```
이 부분도 타이틀 바꿔주세요~

claire-euni commented on students/14th/bjkim/main.html:
```diff
@@ -0,0 +1,96 @@
+          <div class="stories">
+            <img class="toystory" src="img/story.png">
+          </div>
+            <div class="capbox">
+              <div class="boxone">
+                <img class="nmporfile" src="img/row first.png" />
```
capbox, nmporfile은 무슨 의미인지 알기 어려우므로 명확한 클래스명으로 수정해주세요 🙌

claire-euni commented on students/14th/bjkim/main.html:
```diff
@@ -0,0 +1,96 @@
+              </div>
+              <div class="mainlikes">슬퍼요 28개</div>
+              <div class="haha">
+                <span class="chae tears">난 ㄱㅏ... 끔 눈물을 흘린ㄷㅏ...</span>
```
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 아니 싸이월드 갬성

claire-euni reviewed (commented):
병준님 첫 PR 잘 올려주셨네요 👍👍
리뷰반영해서 리팩토링해주세요 ~ 

- 클래스명 명확한 이름으로 수정해주세요. camelCase나 kebab-case식으로 정하셨으면 통일감있게 작성해주세요 :)

joonyg10 committed:
[Add] Login 페이지 레이아웃
- Main페이지 레이아웃 작성중
- Main페이지 feed 이미지 슬라이더 완성
- Main페이지 댓글 달기 로직 작성중

joonyg10 committed:
스토리 슬라이더 구현

joonyg10 committed:
[기능추가] [리펙토링]
좋아요 버튼 클릭시 좋아요수 업데이트 구현
좋아요 버튼 svg이용한 애니메이션 구현
네비게이션바 이미지 클릭시 dropdown리스트 구현
댓글 삭제 기능 구현
자바스크립트, css 정리

joonyg10 committed:
댓글 왼쪽으로 드래그하면 삭제버튼 띄우는 로직 구현
위의 로직 수정 중
각 요소마다 eventListener추가에서 요소를 감싸고 있는 컨테이너에 eventListener를 주어 이벤트 전파하는 형식으로 코드 변경중

joonyg10 committed:
[ADD] 네비게이션의 서치바 입력 시 유저 필터링해서 보여주는 로직 작성

joonyg10 committed:
[Refactoring] 스토리 슬라이더 리펙토링
좋아요 갯수 댓글 하트 눌러도 적용 되도록 수정
댓글창 로직 수정 예정
css refactoring 예정

joonyg10 committed:
댓글 input shift+enter 같이 사용할 때 댓댓글 추가 대신 엔터로 변경
위의 로직 shift+enter시 input의 높이 auto-resize 구현

joonyg10 committed:
[리펙토링] html, css 네비게이션 바 리펙토링
[리펙토링] html, css 스토리 부분 리펙토링

joonyg10 committed:
[리펙토링] 사이드바 html, css 리펙토링

joonyg10 committed:
[Add] feed이미지 슬라이더 밑에 위치하는 navigator html, css, js로직 작성

joonyg10 committed:
[리펙토링] 댓글 왼쪽으로 드래그 할 때, reset, 삭제 버튼 띄우는 부분 리펙토링
댓글부분의 css, js 리펙토링 및 수정

joonyg10 committed:
[ADD] 이미지 위에 skeleton 로딩 효과 추가

joonyg10 committed:
[리뷰요청] main.js의 300줄 target이 undefined로 출력되는 문제
